magento-research/pwa-studio
# bug: Make css-font non render-blocking

## Description
This is bad:
[![Image from Gyazo](https://i.gyazo.com/46bb75cbb34c0b88409b22121240287a.png)](https://gyazo.com/46bb75cbb34c0b88409b22121240287a)

I hardcoded [the response](https://fonts.googleapis.com/css?family=Muli:300) from Google's font API into our mustache template to avoid the render blocking `<link>` inclusion. In the template I also include a new link with the `rel="preload"` attribute for the font that is requested later in the file which tells the browser to start the request for the font ASAP instead of waiting for the CSSOM (where it would normally find the font url) to be built.

So two things were done here:

1) **Inclusion of link rel="preload" for font**: Browser will request font asap instead of waiting for CSSOM.
2) **Include stylesheet inline**: This is done because CSS is a _render-blocking_ request while font is only a _text-blocking_ request. 

This is a pretty extreme optimization but it does improve the lighthouse scores. The image below should help you visualize the improvement (the css is now included in the `html` response).

![](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/images/font-crp.png)

Check out [this awesome documentation](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/webfont-optimization) for more details about web font optimization.

## Related Issue





Closes #1195 .

## Verification Steps
Run the app through lighthouse and see that it no longer complains about the blocking resource:

[![Image from Gyazo](https://i.gyazo.com/800bf57d92a3b8fd5893e54d9a9104da.png)](https://gyazo.com/800bf57d92a3b8fd5893e54d9a9104da)

## How Have YOU Tested this?
* Used lighthouse and saw it no longer complains
* Ran app in chrome and saw that fonts were loaded and displayed properly (Muli is the venia-font)
* Blocked JS on page and reloaded and saw the no-script displayed properly.
* Opened app in Safari, Chrome, Firefox on Desktop 

## Screenshots / Screen Captures (if appropriate):

_(Mocked up a fancier preload screen to show the font behavior. -- Z)_

### (Current) Linked stylesheet with blocking font
![original font stylesheet](https://user-images.githubusercontent.com/1643758/58360688-c25e8000-7e4f-11e9-8a0e-4ccc88a3a23a.gif)

### Linked stylesheet with `display=swap`
![display swap stylesheet](https://user-images.githubusercontent.com/1643758/58360707-d5715000-7e4f-11e9-94ea-7bc10fdd69b1.gif)

### Inlined stylesheet with `font-display: swap`
![inlined font def](https://user-images.githubusercontent.com/1643758/58360766-1c5f4580-7e50-11e9-9582-3add73a8bd4e.gif)


## Proposed Labels for Change Type/Package




## Checklist:
